lib/rubygems/package/tar_header.rb

```diff
@@ -208,7 +208,7 @@ def update_checksum
   private
 
   def calculate_checksum(header)
-    header.bytes.sum
+    header.sum(0)
```

Comment:
Although I don't remember the spec of tar well, as this value is formatted as `"%06o"`, upper bits should be dropped?

```suggestion
    header.sum
```

Reply:
Let's double check just to be sure we're following spec. [Wikipedia](https://en.wikipedia.org/wiki/Tar_(computing)#:~:text=The%20checksum%20is%20calculated%20by,NUL%20and%20then%20a%20space.) says:

> The [checksum](https://en.wikipedia.org/wiki/Checksum) is calculated by taking the sum of the unsigned byte values of the header record with the eight checksum bytes taken to be ASCII spaces (decimal value 32). It is stored as a six digit octal number with leading zeroes followed by a NUL and then a space. Various implementations do not adhere to this format. In addition, some historic tar implementations treated bytes as signed. Implementations typically calculate the checksum both ways, and treat it as good if either the signed or unsigned sum matches the included checksum.

The chars are unsigned, so the sum of the chars should be the correct number, then it's stored as a 0 padded octal. I reproduced the calculation here using an example gem and it produces the same checksum as when I let the old version calculate the header.

```

Reply:
Thanks for challenging me to think about all this. I've learned a bunch just because of this.

Since the header is limited to 512bytes and they're not all used (about 500 are used and a big portion of them are \0.) I think it's impossible for the checksum to exceed the 6 character octal, so any argument other than 0 runs more code than necessary in `String#sum` (if I understand the admittedly cryptic implementation well enough).

You're completely right that 2**18-1 is perfect for the space (777777 in octal). I just don't think it's necessary.